Certora (#1)
Certora/bgd-forge-template: certora/README.md

# Certora 

This folder provides a fully functional skeleton for integrating with Certora.
There are two configurations:

## Run From the Command Line

To run the checks from the command line, first install the Python SDK:

> `pip install certora-cli`

Reviewer: that's often not enough and we need to add the certora package to PATH. At least according to the last time i reinstalled

Author: maybe add link to full installation guide

Reviewer: done

Set your Certora Key:  

> `export CERTORAKEY=<personal_access_key>` 

Reviewer: isnt this only for a given session of a terminal?
Shouldnt we give instructions to add to .zshenv or something?

Reviewer: done

Then run from this directory:  
`certoraRun ./certora/confs/ghost.conf`  
 
## Files

Folder `specs` contains the specification files: 
1. `ghost.spec` contains basic rules that are not dependent on any interface.

> 1. `moreExmaplesERC20.spec` contains basic examples including ghost, hooks, and multi-contract features.

Reviewer: change the numbering to 2

Reviewer: done

Folder `confs` contains the configuration files for running the Certora Prover:  
1. File `ghost.conf` for running `ghost.spec` on `ghost.sol` file.  
This configuration is suitable for a standalone project, containing:  
    * external calls are assumed to be non-state changing and returning random value on each call.  
    * loops are unrolled three times.   
    * calls to fallbacks are assumed to be side-effect-free and only update the native balance.   

>  1. File `ghostWithERC20s.conf` running `moreExmaplesERC20.spec` on `ghost.sol` with additional ERC20-related contracts.   

Reviewer: change the numbering to 2

Author: for me md is formatting this numbered, but ok change

Reviewer: done

This demonstrates how to setup a project with multi-contract 

## CI Integration

This repository contains a Certora setup in GitHub actions.

> You will need to provide your secret `CERTORAKEY` to GitHub, 

Reviewer 2: ```suggestion
You will need to provide your secret `CERTORAKEY` to GitHub. 
```

> ## Mutation Testing

Reviewer: @nd-certora you said perhaps you want to lose the gambit setup for now because it's not stable?
if so we need to remove this as well

Reviewer 2: It is stable, there was just nothing to mutate on the template file

Reviewer: following @urikirsh reply i resolve the comment

# Certora 

This folder provides a fully functional skeleton for integrating with Certora.
There are two configurations:

## Run From the Command Line

Follow this guide to install the prover and its dependencies:
[Installation Guide](https://docs.certora.com/en/latest/docs/user-guide/getting-started/install.html)

Once installations are set, run from this directory:  
`certoraRun ./certora/confs/ghost.conf`  

## Recommended IDE

We recommended using VSCode with the following extension:
[Certora Verification Language LSP](https://marketplace.visualstudio.com/items?itemName=Certora.evmspec-lsp)
This extension is found in the VSCode extensions/marketplace. It provides syntactic support for the Certora Verification Language (CVL) - the language in which we will be writing specifications.
 
## Files

> Folder `specs` contains the specification files: 

Reviewer 2: ```suggestion
The folder `specs` contains the specification files: 
```

1. `ghost.spec` contains basic rules that are not dependent on any interface.
2. `moreExmaplesERC20.spec` contains basic examples including ghost, hooks, and multi-contract features.

Folder `confs` contains the configuration files for running the Certora Prover:  
1. File `ghost.conf` for running `ghost.spec` on `ghost.sol` file.  
This configuration is suitable for a standalone project, containing:  

>     * external calls are assumed to be non-state changing and returning random value on each call.  

Reviewer 2: ```suggestion
    * external calls are assumed to be non-state changing and return random values on each call.  
```

    * loops are unrolled three times.   
    * calls to fallbacks are assumed to be side-effect-free and only update the native balance.   
2. File `ghostWithERC20s.conf` running `moreExmaplesERC20.spec` on `ghost.sol` with additional ERC20-related contracts.   

> This demonstrates how to setup a project with multi-contract 

Reviewer 2: ```suggestion
This demonstrates how to set up a project with multi-contract 
```

## CI Integration

This repository contains a Certora setup in GitHub actions.
You will need to provide your secret `CERTORAKEY` to GitHub, 

## Mutation Testing
The Certora Prover is integrated with <a href=https://www.certora.com/gambit>Gambit</a>, an open-source Solidity mutation testing tool. Mutation testing can give you an estimate of the coverage of your verification.

To run the mutation test from this directory, run `certoraMutate --prover_conf certora/confs/ghost.conf --mutation_conf certora/confs/mutation.mconf`

## CVL Examples and Docs
See more <a href="https://github.com/Certora/Examples/tree/master/CVLByExample" target="_blank">CVL specification general examples</a>

> and  <a href="https://docs.certora.com" target="_blank">Documenation</a>.

Reviewer 2: ```suggestion
and <a href="https://docs.certora.com" target="_blank">Documenation</a>.
```
